---
title: '00A - Sequencing read pre-processing'
author: "Emma Bell"
date: '2020-04-22'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary {.tabset}

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020).

* Aim: To pre-process human sequencing read data.
* Objectives:
  1. Assess the quality of the sequencing reads using FastQC;
  2. Trim sequencing adaptors, base pairs with a Phred score < 30, and reads with a length < 20 bp;
  3. Align the reads using Bowtie.[^1]
  
[^1] I aligned the sequencing reads from the ATAC-seq and RNA-seq experiments with Bowtie. For the RRBS experiment, I used Bismark, a wrapper for Bowtie that enables alignment of bisulphite-converted DNA. See: RRBS_01 - Sequencing read alignment with Bismark.

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * RRBS_04 - Defining hypomethylated windows for HOMER motif enrichment
  * ATAC_05 - Calling nucleosome-free regions with MACS2
  * ATAC_06 - Quantifying nucleosome-free regions
  * ATAC_07 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNA_08 - Quantifying transcript expression with Kallisto
  * RNA_09 - Calling differential transcript expression with Sleuth
  * RNA_10 - Gene-set enrichment with CensensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/

## 1. Assess the quality of the sequencing reads using FastQC

I ran FastQC as follows:

```{bash eval=FALSE}
module load fastqc/0.11.5

ls -1 data/01_raw/fastq/*.fastq > data/01_raw/fastqs.txt

mkdir data/01_raw/fastqc

while read FASTQ;
do
fastqc $FASTQ --outdir data/01_raw/fastqc
done < data/01_raw/fastqs.txt
```

## 2. Trim sequencing adaptors, base pairs with a Phred score < 30, and reads with a length < 20 bp

## 3. Align the reads using Bowtie

