---
title: 'ATAC_05 - Characterising the nucleosome-free regions'
author: "Emma Bell"
date: '2020-05-01'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020) _Nature Medicine_.

* Aim: To characterise the nucleosome-free regions.
* Objectives:
  1. To visualise the number of nucleosome-free regions in each sample; 
  2. To identify nucleosome-free regions common to all samples within each treatment group;
  3. To determine the overlap in nucleosome-free regions between treatment groups.

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * ATAC_04 - Calling nucleosome-free regions with MACS2
  * ATAC_05 - Quantifying nucleosome-free regions
  * ATAC_06 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNA_07 - Quantifying transcript expression with Kallisto
  * RNA_08 - Calling differential transcript expression with Sleuth
  * RNA_09 - Gene-set enrichment with CensensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/

## Configuration

This analysis requires the following packages:

```{r}
library(readxl)
library(vtree)
library(GenomicRanges)
library(BiocParallel)
library(ChIPpeakAnno)
library(UpSetR)
library(writexl)
```

I use the `BiocParallel` package to parallelise R with the following parameters:

```{r}
param <- MulticoreParam(workers = 10, progressbar = TRUE)
```

This analysis requires the following R configuration object:

```{r}
r.config <- readRDS(file = file.path("code","r_config.RDS"))
```

## Metadata

```{r}
metadata <- read_excel(path = file.path(r.config$dirs$data.dir, "metadata.xlsx"), sheet = "ATAC")
metadata
```

I'll use the `vtree` package to visualise the experimental design.

```{r}
vtree(z = metadata, vars = c("Stimulated","DAC"))
```

# Analysis

## 1. Visualising the number of nucleosome-free regions in each sample

```{r}
atac.peaks <- bplapply(X = metadata$File, FUN = function(X){
  to.read <- X
  this.result <- read.table(file = file.path(r.config$dirs$augmented.data.dir,"03_atac_nfrs","01_narrowPeaks",to.read), stringsAsFactors = FALSE)
  this.result.gr <- GRanges(seqnames = this.result[,1], ranges = IRanges(start = this.result[,2], end = this.result[,3]))
  mcols(this.result.gr) <- cbind.data.frame(read.counts = this.result[,4])
  this.result.gr
}, BPPARAM = param)
```

```{r}
num.peaks <- sapply(X = atac.peaks, FUN = length)
num.peaks <- matrix(data = num.peaks, nrow = 3, ncol = 3, byrow = FALSE)
colnames(num.peaks) <- paste("Donor",unique(metadata$Donor))
rownames(num.peaks) <- c("Unstimulated","Stimulated","DAC")
num.peaks
```

```{r}
ylab <- "# NFRs (thousands)"
barplot(height = num.peaks/1000, beside = TRUE, col = r.config$simple.cols, las = 1, ylab = ylab)
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"ATAC_S01_NFRbarplot.png")
png(filename = png.name, width = 4, height = 4, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
par(mar = c(1,4,1,1))
barplot(height = num.peaks/1000, beside = TRUE, col = r.config$simple.cols, las = 1, ylab = ylab)
dev.off()
```

## 2. Identifying nucleosome-free regions common to all samples within each treatment group

```{r}
treatment.groups <- factor(paste(metadata$Stimulated,metadata$DAC, sep = "_"), levels = c("Unstimulated_Untreated","Stimulated_Untreated","Stimulated_Treated"))
treatment.groups
```


```{r}
atac.groups <- bplapply(X = levels(treatment.groups), FUN = function(X){
  this.group <- X
  group.index <- which(treatment.groups == this.group)
  subset1 <- subsetByOverlaps(x = atac.peaks[[group.index[1]]], ranges = atac.peaks[[group.index[2]]])
  subset2 <- subsetByOverlaps(x = subset1, ranges = atac.peaks[[group.index[3]]])
  subset2
}, BPPARAM = param)
names(atac.groups) <- levels(treatment.groups)
```

```{r}
saveRDS(object = atac.groups, file = file.path(r.config$dirs$augmented.data.dir,"03_atac_nfrs","atac_nfrs_byGroup.RDS"))
```


```{r}
num.common.peaks <- sapply(X = atac.groups, FUN = length)
num.common.peaks
```

```{r}
labels <- prettyNum(x = num.common.peaks, big.mark=",")
ylab <- "# Common NFRs (thousands)"
ylim <- c(0,max(num.common.peaks/1000) * 1.1)
bp.x <- barplot(height = num.common.peaks, plot = FALSE) 
barplot(height = num.common.peaks/1000, col = r.config$simple.cols, ylim = ylim, names.arg = NA, las = 1, ylab = ylab)
text(x = bp.x, y = num.common.peaks/1000 + (ylim[2] * 0.05), labels = labels)
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"ATAC_01_commonNFR.png")
png(filename = png.name, width = 4, height = 4, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
par(mar = c(1,4,1,1))
barplot(height = num.common.peaks/1000, col = r.config$simple.cols, ylim = ylim, names.arg = NA, las = 1, ylab = ylab)
text(x = bp.x, y = num.common.peaks/1000 + (ylim[2] * 0.05), labels = labels)
dev.off()
```

```{r}
atac.df <- bplapply(X = atac.groups, FUN = function(X){
  these.peaks <- X
  chr <- seqnames(these.peaks)
  start <- start(these.peaks)
  end <- end(these.peaks)
  peak.id <- mcols(these.peaks)$read.counts
  data.frame(chr=chr,start=start,end=end,peak.id=peak.id)
}, BPPARAM = param)
```


```{r}
write_xlsx(x = atac.df, path = file.path(r.config$dirs$table.dir,"ATAC_SupplementaryTable_01.xlsx"), col_names = TRUE)
```


## 3. Determining the overlap in nucleosome-free regions between treatment groups

```{r}
counts <- makeVennDiagram(Peaks = atac.groups)
counts
```

```{r}
counts.expression <- c(Unstimulated = counts$vennCounts[5,4], Stimulated = counts$vennCounts[3,4], `Stimulated+DAC` = counts$vennCounts[2,4], `Unstimulated&Stimulated` = counts$vennCounts[7,4], `Unstimulated&Stimulated+DAC` = counts$vennCounts[6,4], `Stimulated&Stimulated+DAC` = counts$vennCounts[4,4], `Unstimulated&Stimulated&Stimulated+DAC` = counts$vennCounts[8,4])
names(counts.expression) <- gsub(pattern = ".Counts", replacement = "", x = names(counts.expression))
counts.expression
```


```{r}
upset(fromExpression(counts.expression), sets.bar.color = r.config$simple.cols, sets = c("Unstimulated","Stimulated","Stimulated+DAC"), keep.order = TRUE)
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"ATAC_02_NFRupset.png")
png(filename = png.name, width = 10, height = 10, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
upset(fromExpression(counts.expression), sets.bar.color = r.config$simple.cols, sets = c("Unstimulated","Stimulated","Stimulated+DAC"), keep.order = TRUE, text.scale = 0.2)
dev.off()
```
