---
title: 'ATAC_04 - Calling nucleosome-free regions with MACS2'
author: "Emma Bell"
date: '2020-04-30'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020) _Nature Medicine_.

* Aim: To call nucleosome-free regions with MACS2.
* Objectives:
  1. To use MACS2 to call nucleosome-free regions.

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * ATAC_04 - Calling nucleosome-free regions with MACS2
  * ATAC_05 - Quantifying nucleosome-free regions
  * ATAC_06 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNA_07 - Quantifying transcript expression with Kallisto
  * RNA_08 - Calling differential transcript expression with Sleuth
  * RNA_09 - Gene-set enrichment with CensensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/

## Configuration

This analysis requires the following packages:

```{r}
library(readxl)
library(vtree)
library(writexl)
```

This analysis requires the following R configuration object:

```{r}
r.config <- readRDS(file = file.path("code","r_config.RDS"))
```

## Metadata

```{r}
metadata <- read_excel(path = file.path(r.config$dirs$data.dir, "metadata.xlsx"), sheet = "ATAC")
metadata
```

I'll use the `vtree` package to visualise the experimental design.

```{r}
vtree(z = metadata, vars = c("Stimulated","DAC"))
```

# Analysis

## 1. Using MACS2 to call nucleosome-free regions

These are the arguments to MACS2 we used to call nucleosome-free regions.

```{bash eval=FALSE}
macs2 callpeak -t $BAM -g hs -n $OUTNAME -q 0.25 --nomodel --shift -100 --extsize 200
```

The resulting narrowPeak files are located at `data/03_augmented/03_atac_nfrs/01_narrowPeaks`.
