---
title: 'RNAseq_07 - Quantifying transcript abundance and differential expression with Kallisto and Sleuth'
author: "Emma Bell"
date: '2020-05-04'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020) _Nature Medicine_.

* Aim: To quantify transcript expression with Kallisto.
* Objectives:
  1. To use Kallisto to get transcript abundances for each sample;
  2. To use Sleuth to call differential expression;
  3. To compile tables of estimated counts and TPM;
  4. To perform Principal Components Analysis;
  5. To get lists of the top genes driving PC1 and PC2;
  6. To visualise the differentially expressed transcripts at a heatmap.

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * ATAC_04 - Calling nucleosome-free regions with MACS2
  * ATAC_05 - Quantifying nucleosome-free regions
  * ATAC_06 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNA_07 - Quantifying transcript abundance and differential expression with Kallisto and Sleuth
  * RNA_08 - Gene-set enrichment with CensensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/

## Configuration

This analysis requires the following packages:

```{r}
library(readxl)
library(vtree)
library(sleuth)
library(writexl)
library(pheatmap)
```

This analysis requires the following R configuration object:

```{r}
r.config <- readRDS(file = file.path("code","r_config.RDS"))
```

## Metadata

```{r}
metadata <- read_excel(path = file.path(r.config$dirs$data.dir, "metadata.xlsx"), sheet = "RNAseq")
metadata
```

I'll use the `vtree` package to visualise the experimental design.

```{r}
vtree(z = metadata, vars = c("Stimulated","DAC"))
```

# Analysis

## 1. Using kallisto quant to get transcript abundances for each sample

These are the arguments to Kallisto Quant we used to get transcript abundances.

```{bash eval=FALSE}
kallisto quant -b 100 -i hg19_kindex -o $OUTDIR -t 18 --bias $FASTQ_R1 $FASTQ_R2
```

The resulting folders containing the transcript abundance files are located at `data/03_augmented/05_rnaseq_kallisto/01_kallisto_quant`.

## 2. Using Sleuth to call differential expression

```{r}
file.paths <- list.files(path = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","01_kallisto_quant"), pattern="Donor")
file.paths <- file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","01_kallisto_quant",file.paths)

s2c <- cbind.data.frame(metadata, path = file.paths, stringsAsFactors = FALSE)
colnames(s2c)[1] <- "sample"
s2c$Stimulated <- c(0,1,1)
s2c$DAC <- c(0,0,1)

so <- sleuth_prep(sample_to_covariates = s2c, ~ Donor + Stimulated + DAC)
```

```{r}
so <- sleuth_fit(so, ~Donor + Stimulated, 'reduced')
so <- sleuth_fit(so, ~Donor + Stimulated + DAC, "full")
so <- sleuth_lrt(so, 'reduced', 'full')
sleuth.de <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE, pval_aggregate = FALSE)
```

```{r}
length(which(sleuth.de$qval<=0.05))
```

```{bash eval=FALSE}
mkdir data/03_augmented/05_rnaseq_kallisto/02_sleuth
```


```{r}
write.table(x = sleuth.de, file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuthDE_DACvsStim.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
sleuth.signif <- sleuth.de[which(sleuth.de$qval<=0.05),]
write.table(x = sleuth.signif, file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuthDE_DACvsStim_signif.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
```

```{r}
write_xlsx(x = sleuth.signif, path = file.path(r.config$dirs$table.dir,"RNAseq_SupplementaryTable_01.xlsx"), col_names = TRUE, format_headers = TRUE)
```

## 3. Compiling tables of estimated counts and TPM

```{r}
tpm <- sleuth_to_matrix(obj = so, which_df = 'obs_norm', which_units = 'tpm')
write.table(x = tpm, file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuth_tpm.txt"), sep = "\t", quote = FALSE)
counts <- sleuth_to_matrix(obj = so, which_df = 'obs_norm', which_units = "est_counts")
write.table(x = counts, file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuth_estCounts.txt"), sep = "\t", quote = FALSE)
```

## 4. Performing Principal Components Analysis

```{r}
pc <- prcomp(x = t(counts))
var <- round(x = 100 * (pc$sdev/sum(pc$sdev)), digits = 1)
axis.names <- paste0("PC ",1:length(var)," (",var,"%)")
xlim <- range(pc$x[,1]) * 1.1
ylim <- range(pc$x[,2]) * 1.1
plot(x = pc$x[,1], y = pc$x[,2], col = rep(r.config$simple.cols,3), pch = 19, xlab = axis.names[1], ylab = axis.names[2])
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"RNAseq_01_PCA.png")
png(filename = png.name, width = r.config$fig.specs$width, height = r.config$fig.specs$height, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
par(mar = c(4,4,2,2))
plot(x = pc$x[,1], y = pc$x[,2], col = rep(r.config$simple.cols,3), pch = 19, xlab = axis.names[1], ylab = axis.names[2], xlim = xlim, ylim = ylim)
dev.off()
```

## 5. Getting lists of the top genes driving PC1 and PC2

```{r}
rotation <- pc$rotation[,1]
# rotation.abs <- abs(rotation)
rotation.abs <- sort(rotation,decreasing = T)
drivers <- names(rotation.abs)[1:10]
driver.genes <- sapply(drivers,function(x) strsplit(x,split = "\\|")[[1]][6])
barplot(height = rotation.abs[1:10], names = driver.genes, las=2 , cex.names = 0.8, main = axis.names[1])
```

```{r}
rotation <- pc$rotation[,2]
# rotation.abs <- abs(rotation)
rotation.abs <- sort(rotation,decreasing = T)
drivers <- names(rotation.abs)[1:10]
driver.genes <- sapply(drivers,function(x) strsplit(x,split = "\\|")[[1]][6])
barplot(height = rotation.abs[1:10], names = driver.genes, las=2 , cex.names = 0.8, main = axis.names[2])
```

## 6. Visualising the differentially expressed transcripts at a heatmap.

```{r}
signif <- sleuth.de$target_id[which(sleuth.de$qval <= 0.05)]
annotation_col <- data.frame(metadata[,3:5])
rownames(annotation_col) <- colnames(tpm)
```

```{r}
pheatmap(mat = tpm[signif,], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = FALSE, scale = "row", clustering_method = "ward.D2", annotation_colors = r.config$fig.cols, annotation_col = annotation_col, show_colnames = FALSE, treeheight_row = 15, treeheight_col = 15, fontsize = 6)
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"RNAseq_02_Heatmap.png")
png(filename = png.name, width = 8, height = r.config$fig.specs$height, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
pheatmap(mat = tpm[signif,], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = FALSE, scale = "row", clustering_method = "ward.D2", annotation_colors = r.config$fig.cols, annotation_col = annotation_col, show_colnames = FALSE, treeheight_row = 15, treeheight_col = 15, fontsize = 6)
dev.off()
```


