---
title: 'RNAseq_08 - Gene-set enrichment with ConsensusPathDB and limma'
author: "Emma Bell"
date: '2020-05-04'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020) _Nature Medicine_.

* Aim: To perform gene-set enrichment with ConsensusPathDB and limma
* Objectives:
  1. To perform gene-set enrichment;
  2. To visualise the differential transcript expression of candidate pathways;
  3. To visualise the NFAT locus;
  4. To plot the expression of NFAT transcripts;
  5. To plot the expression of NFAT related pathways.

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * ATAC_04 - Calling nucleosome-free regions with MACS2
  * ATAC_05 - Quantifying nucleosome-free regions
  * ATAC_06 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNAseq_07 - Quantifying transcript abundance and differential expression with Kallisto and Sleuth
  * RNAseq_08 - Gene-set enrichment with ConsensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/

## Configuration

This analysis requires the following packages:

```{r}
library(readxl)
library(vtree)
library(BiocParallel)
library(limma)
library(writexl)
library(pheatmap)
library(EnsDb.Hsapiens.v75)
library(Gviz)
library(viridis)
library(vioplot)
```

I use the `BiocParallel` package to parallelise R with the following parameters:

```{r}
param <- MulticoreParam(workers = 10, progressbar = TRUE)
```

This analysis requires the following R configuration object:

```{r}
r.config <- readRDS(file = file.path("code","r_config.RDS"))
```

## Metadata

```{r}
metadata <- read_excel(path = file.path(r.config$dirs$data.dir, "metadata.xlsx"), sheet = "RNAseq")
metadata
```

I'll use the `vtree` package to visualise the experimental design.

```{r}
vtree(z = metadata, vars = c("Stimulated","DAC"))
```

# Analysis

## 1. Performing gene-set enrichment

For gene-set enrichment I use a list of pathways downloaded from ConsensusPathDB.

```{r}
sleuth.de <- read.table(file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuthDE_DACvsStim.txt"), header = TRUE, sep = "\t", stringsAsFactors = FALSE)
pathways <- read.table(file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","04_geneSetEnrichment","CPDB_pathways_genes.tab"), header = TRUE, sep = "\t", stringsAsFactors = FALSE, comment.char = "")
```

```{r}
gene.names <- sapply(X = sleuth.de$target_id, FUN = function(X) strsplit(x = X, split = "\\|")[[1]][6])
```


```{r}
pathway.genes <- bplapply(X = seq_along(1:nrow(pathways)), FUN = function(i) strsplit(as.character(pathways$hgnc_symbol_ids[i]),split=",")[[1]], BPPARAM = param)
names(pathway.genes) <- pathways$pathway
```

```{r eval=FALSE}
pathway.enrichments <- bplapply(X = pathway.genes, FUN = function(X){
  geneSetTest(index = which(gene.names %in% X), statistics = sleuth.de$test_stat, ranks.only = FALSE, type = "f", alternative = "mixed")
}, BPPARAM = param)
pathway.enrichments.df <- cbind.data.frame(pathways, pval = unlist(pathway.enrichments), stringsAsFactors = FALSE)
pathway.enrichments.df$adjPVal <- p.adjust(pathway.enrichments.df$pval, method = "BH")
pathway.enrichments.df$NegLog10AdjPVal <- -1 * log10(pathway.enrichments.df$adjPVal)
pathway.enrichments.df <- pathway.enrichments.df[order(pathway.enrichments.df$adjPVal),]
write.table(x = pathway.enrichments.df, file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","04_geneSetEnrichment","pathwayEnrichments.txt"), sep = "\t", quote = FALSE, row.names = FALSE)
write_xlsx(x = pathway.enrichments.df, path = file.path(r.config$dirs$table.dir,"RNAseq_SupplementaryTable_03.xlsx"), col_names = TRUE, format_headers = TRUE)
```

```{r}
pathway.enrichments.xlsx <- read_excel(path = file.path(r.config$dirs$table.dir,"RNAseq_SupplementaryTable_03.xlsx"))
classesToPlot <- c("Calcium","MAPK","NFAT","PI3K","Wnt")
pathways.toPlot <- pathway.enrichments.xlsx[which(pathway.enrichments.xlsx$adjPVal <= 0.05 & pathway.enrichments.xlsx$Class %in% classesToPlot),]
pathways.toPlot <- pathways.toPlot[order(pathways.toPlot$Class, decreasing = TRUE),]
pathways.toPlot
```

```{r}
write_xlsx(x = pathways.toPlot, path = file.path(r.config$dirs$figure.dir,"RNAseq_S01_pathwayEnrichment.xlsx"), col_names = TRUE, format_headers = TRUE)
```


```{r}
xlab <- expression(paste("-Log"[10],"(Adj. P-value)"))
par(mar = c(4,1,1,1))
barplot(height = pathways.toPlot$negLog10AdjP, horiz = TRUE, xlab = xlab)
abline(v = (-1*log10(0.05)), lwd = 2, lty = 2)
```

```{r}
png.name <- file.path(r.config$dirs$figure.dir,"RNAseq_03_pathwayEnrichment.png")
png(filename = png.name, width = 3, height = r.config$fig.specs$height, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
par(mar = c(4,1,1,1))
barplot(height = pathways.toPlot$negLog10AdjP, horiz = TRUE, xlab = xlab)
abline(v = (-1*log10(0.05)), lwd = 2, lty = 2)
dev.off()
```

## 3. Visualising the NFAT locus


```{r}
ens.db <- EnsDb.Hsapiens.v75
nfat.ens <- "ENSG00000131196"
gf <- GeneNameFilter("NFATC1")
tbtf <- TxBiotypeFilter("protein_coding", condition = "==")
```

```{r}
tpm <- read.table(file = file.path(r.config$dirs$augmented.data.dir,"05_rnaseq_kallisto","02_sleuth","sleuth_tpm.txt"), sep = "\t", header = TRUE, stringsAsFactors = FALSE)
```

```{r}
index <- grep(pattern = nfat.ens, x = rownames(tpm))
nfat.annot <- data.frame(t(data.frame(lapply(X = rownames(tpm)[index], FUN = function(X) strsplit(X, split = "\\|")[[1]][1:6]), stringsAsFactors = FALSE)), stringsAsFactors = FALSE)
colnames(nfat.annot) <- c("tx_id","gene_id","ott_gene","ott_tx","tx_name","gene_name")
nfat.annot$tx_id <- substr(x = nfat.annot$tx_id, start = 1, stop = nchar(nfat.annot$tx_id) - 2)
```

```{r}
nfat.tx <- transcripts(x = ens.db, filter = AnnotationFilterList(gf, tbtf),  order.by = "tx_seq_start")
nfat.exons <- exons(x = ens.db, filter = AnnotationFilterList(gf, tbtf), columns = c("gene_id", "tx_id","gene_seq_start", "gene_seq_end", "tx_seq_start", "tx_cds_seq_start", "tx_cds_seq_end", "tx_biotype"), order.by = "tx_seq_start")
nfat.exons.df <- data.frame(chromosome = paste0("chr",seqnames(nfat.exons)), start = start(nfat.exons), end = end(nfat.exons), width = width(nfat.exons), strand = "+", feature = "protein_coding", gene = nfat.exons$gene_id, exon = nfat.exons$exon_id, transcript = nfat.exons$tx_id)
nfat.exons.df <- merge(nfat.exons.df, nfat.annot[,c("tx_id","tx_name")], by.x = "transcript", by.y = "tx_id")
colnames(nfat.exons.df)[10] <- "symbol"
```

```{r}
chr <- "chr18"
gen <- "hg19"
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome=gen, chromosome=chr)
grtrack <- GeneRegionTrack(range = nfat.exons.df, genome = "hg19", chromosome = "chr18", name="NFATC1", strand = "+", col = "darkgrey", fill = "grey")
```

```{r}
plotTracks(list(itrack,gtrack,grtrack), chromosome = 18, from = 77090000, to = 77289325, extend.right = 0.05, transcriptAnnotation="symbol")
```


```{r}
png.name <- file.path(r.config$dirs$figure.dir,"RNAseq_S02_NFATlocus.png")
png(filename = png.name, width = 15, height = 8, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = 12)
plotTracks(list(itrack,gtrack,grtrack), chromosome = 18, from = 77090000, to = 77289325, extend.right = 0.05, transcriptAnnotation="symbol")
dev.off()
```

## 4. Plotting the expression of NFAT transcripts

```{r}
index.sleuth <- grep(pattern = nfat.ens, x = sleuth.de$target_id)
sleuth.de[index.sleuth,]
```

```{r}
index <- which(grepl(pattern = nfat.ens, x = rownames(tpm)) & rowSums(tpm) > 0)
nfat.tpm <- tpm[index,]
tx.names <- rownames(nfat.tpm)
tx.names <- sapply(X = tx.names, FUN = function(X) strsplit(X, split = "\\|")[[1]][5])
rownames(nfat.tpm) <- tx.names
nfat.tpm <- nfat.tpm[order(rownames(nfat.tpm)),]
```

```{r}
annotation_col <- data.frame(metadata[,3:5], stringsAsFactors = FALSE)
rownames(annotation_col) <- colnames(tpm)
colnames(annotation_col)[1] <- "Donor"
```

```{r}
pheatmap(mat = nfat.tpm, color = viridis(n = 100), clustering_method = "ward.D2", border_color = NA, annotation_col = annotation_col, annotation_colors = r.config$fig.cols, show_colnames = FALSE)
```

```{r}
png.name <- file.path(r.config$dirs$figure.dir,"RNAseq_04_NFATheatmap.png")
png(filename = png.name, width = 15, height = 10, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
pheatmap(mat = nfat.tpm, color = viridis(n = 100), clustering_method = "ward.D2", border_color = NA, annotation_col = annotation_col, annotation_colors = r.config$fig.cols, show_colnames = FALSE, fontsize = 8)
dev.off()
```

```{r}
nfat.pvals <- sleuth.de[index.sleuth,c(1:2)]
qvals <- data.frame(q = p.adjust(p = nfat.pvals$pval, method = "BH"), stringsAsFactors = FALSE)
qval.rownames <- sapply(X = nfat.pvals$target_id, FUN = function(X) strsplit(X, split = "\\|")[[1]][5])
rownames(qvals) <- qval.rownames
qvals
```

```{r}
nfats <- rownames(nfat.tpm)
nfat.data <- unlist(nfat.tpm)
nfat.donor <- rep(c(1,2,3), each = length(nfat.data)/3)
nfat.treatment <- rep(c("Unstimulated","Stimulated","Stimulated+DAC"), each = nrow(nfat.tpm))
nfat.tpm.df <- data.frame(donor = nfat.donor, treatment = factor(nfat.treatment, levels = c("Unstimulated","Stimulated","Stimulated+DAC")), transcript = rownames(nfat.tpm), data = nfat.data)
```


```{r}
for(i in 1:length(nfats)){
  this.nfat <- nfats[i]
  this.data <- nfat.tpm.df[grep(pattern = this.nfat, x = nfat.tpm.df$transcript),]
  ylim <- c(0, max(this.data$data) * 1.2)
  par(mar = c(1,4,1,1))
  boxplot(data ~ treatment, data = this.data, type="n", ann = TRUE, col = "white", border = "white", ylab = "TPM", xlab = "", pch = 20, ylim = ylim, las = 1, xaxt = "n")
  axis(side = 1, tick = FALSE, labels = FALSE)
  if(this.nfat %in% rownames(qvals)){
    qval <- round(x = qvals[this.nfat,1], digits = 3)
    text(x = 2.5, y = ylim[2] * 0.95, labels = paste0("q = ",qval))
    lines(x = c(2,3), y = rep((ylim[2] * 0.9),2))
  }
  vioplot(lapply(X = 1:3, FUN = function(X){ this.data[which(as.numeric(this.data$treatment) == X),"data"]}), col = "white", border = r.config$simple.cols, add = TRUE, rectCol = r.config$simple.cols, pchMed = "—", colMed = r.config$simple.cols, lineCol = r.config$simple.cols, lwd = 2)
  stripchart(data ~ treatment, data = this.data, vertical = TRUE, method = "jitter", pch = 20, cex = 1.5, add = TRUE, col = r.config$simple.cols)
}
```


```{r}
for(i in 1:length(nfats)){
  this.nfat <- nfats[i]
  this.data <- nfat.tpm.df[grep(pattern = this.nfat, x = nfat.tpm.df$transcript),]
  ylim <- c(0, max(this.data$data) * 1.2)
  file.name <- paste0("RNAseq_05_",this.nfat,"_vioplot.png")
  png.name <- file.path(r.config$dirs$figure.dir,file.name)
  png(filename = png.name, width = 4, height = 4, units = r.config$fig.specs$units, res = r.config$fig.specs$res, pointsize = r.config$fig.specs$pointsize)
  par(mar = c(1,4,1,1))
  boxplot(data ~ treatment, data = this.data, type="n", ann = TRUE, col = "white", border = "white", ylab = "TPM", xlab = "", pch = 20, ylim = ylim, las = 1, xaxt = "n")
  axis(side = 1, tick = FALSE, labels = FALSE)
  if(this.nfat %in% rownames(qvals)){
    qval <- round(x = qvals[this.nfat,1], digits = 3)
    text(x = 2.5, y = ylim[2] * 0.95, labels = paste0("q = ",qval))
    lines(x = c(2,3), y = rep((ylim[2] * 0.9),2))
  }
  vioplot(lapply(X = 1:3, FUN = function(X){ this.data[which(as.numeric(this.data$treatment) == X),"data"]}), col = "white", border = r.config$simple.cols, add = TRUE, rectCol = r.config$simple.cols, pchMed = "—", colMed = r.config$simple.cols, lineCol = r.config$simple.cols, lwd = 2)
  stripchart(data ~ treatment, data = this.data, vertical = TRUE, method = "jitter", pch = 20, cex = 1.5, add = TRUE, col = r.config$simple.cols)
  dev.off()
}
```
