---
title: 'RRBS_01 - Sequencing read alignment with Bismark'
author: "Emma Bell"
date: '2020-04-27'
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
    toc_depth: 2
    fig_caption: true
---

# Summary

This R Notebook contains code pertaining to the paper Loo and Bell _et al._ (2020) _Nature Medicine_.

* Aim: To align the RRBS sequencing reads using Bismark.
* Objectives:
  1. Trim fastq files using TrimGalore; 
  2. Align to a bisulphite converted genome using Bismark;
  3. Sort and index bam files using Samtools;
  4. Create bedgraph files using bismark_methylation_extractor.
  
I adapted these scripts from a RRBS pipeline written by Roxana Shen (version 2018-03-06).

To run these scripts we assume a directory structure as described in the README accompanying this GitHub. I.e.

* code/ - _contains R Notebook(s) and the resulting HTML file(s)_
  * 00A - Sequencing read pre-processing
  * 00B - R configuration
  * RRBS_01 - Sequencing read alignment with Bismark
  * RRBS_02 - Calling base level methylation with MethylKit
  * RRBS_03 - Calculating differential methylation with MethylKit
  * RRBS_04 - Defining hypomethylated windows for HOMER motif enrichment
  * ATAC_05 - Calling nucleosome-free regions with MACS2
  * ATAC_06 - Quantifying nucleosome-free regions
  * ATAC_07 - Defining nucleosome-free regions unique to the DAC sample group for HOMER motif enrichment
  * RNA_08 - Quantifying transcript expression with Kallisto
  * RNA_09 - Calling differential transcript expression with Sleuth
  * RNA_10 - Gene-set enrichment with CensensusPathDB and limma
* data/ - _contains any data used in this project_
  * 01_raw/
  * 02_clean/
  * 03_augmented/
* docs/ - _contains any documents pertaining to this project_
  * Presentation_plan_template.docx
  * Presentation_template.pptx
  * Research_project_plan.docx
* results/ - _contains results files_
  * docs/
  * figures/
  * tables/
  
# Analysis

## 1. Trimming fastq files using TrimGalore

```{bash eval=FALSE}
echo "Running RRBS analysis pipeline..."
mkdir data/02_clean/01_rrbs_trimmedFastqs
mkdir data/02_clean/02_rrbs_bams
mkdir data/02_clean/03_rrbs_bedgraphs
```

Remove the two bases pertaining to the cytosine artificially introduced during the end-repair step. Trim Galore simultaneously runs FastQC to evaluate the quality of the trimmed fastq files.

```{bash eval=FALSE}
echo "Trimming fastq files..."
module load trim_galore/0.5.0
module load fastqc/0.11.5
find data/ -name "*.fastq.gz" -print0 | while IFS= read -r -d '' FASTQ; 
do
  trim_galore --quality 20 --fastqc_args "--nogroup --outdir data/02_clean/01_rrbs_trimmedFastqs" --stringency 5 --length 20 --output_dir data/02_clean/01_rrbs_trimmedFastqs --rrbs $FASTQ
done
echo "Trimming fastq files... completed."
```

## 2. Aligning to the bisulphite converted human genome using Bismark;

```{bash eval=FALSE}
echo "Aligning trimmed fastq files..."
module load bowtie/1.2.1.1 #version used as of Jan-31-2018: 1.2.1.1
module load bismark/0.16.3
module load samtools/1.8

find data/ -name "*_trimmed.fq.gz" -print0 | while IFS= read -r -d '' TRIMMED_FASTQ; 
do
  echo $TRIMMED_FASTQ
  bismark --fastq --bowtie1 --bam --output_dir data/02_clean/02_rrbs_bams --genome_folder /mnt/work1/data/commondata/bismark/hg19 --single_end $TRIMMED_FASTQ
done
echo "Aligning trimmed fastq files... completed."
```

## 3. Sorting and indexing bam files using Samtools

```{bash eval=FALSE}
echo "Sorting and indexing bam files..."
find data/ -name "*.bam" -print0 | while IFS= read -r -d '' BAM; 
do
  echo $BAM
  BAMNAME=$(echo $BAM | cut -d "/" -f 4)
  samtools sort -@ 12 -o data/02_clean/02_rrbs_bams/$BAMNAME $BAM
  samtools index -b $BAM data/02_clean/02_rrbs_bams/$BAMNAME".bai"
done
echo "Sorting and indexing bam files... completed."
```

## 4. Create bedgraph files using bismark_methylation_extractor

Replace the argument to `--genome_folder` with your local path to the Bismark processed version of the human genome.

```{bash eval=FALSE}
echo "Extracting methylation values and generating bedGraph..."
find data/ -name "*.bam" -print0 | while IFS= read -r -d '' BAM; 
do
  bismark_methylation_extractor -s --output data/02_clean/03_rrbs_bedgraphs --bedgraph --genome_folder /mnt/work1/data/commondata/bismark/hg19 $BAM
done
echo "Extracting methylation values and generating bedGraph... completed."

echo "Running RRBS analysis pipeline... completed."
```
